lib/amazonka-core/src/Amazonka/Sign/V4/Base.hs

```diff
@@ -271,19 +274,18 @@ canonicalQuery = Tag . toBS
 -- all internal whitespace, replacing with a single space char,
 -- unless quoted with \"...\"
 canonicalHeaders :: NormalisedHeaders -> CanonicalHeaders
-canonicalHeaders = Tag . Foldable.foldMap (uncurry f) . untag
+canonicalHeaders = Tag . BSL.toStrict . BSB.toLazyByteString . Foldable.foldMap (uncurry f) . untag
   where
-    f k v = k <> ":" <> stripBS v <> "\n"
+    f k v = BSB.byteString k <> BSB.char7 ':' <> BSB.byteString (stripBS v) <> BSB.char7 '\n'
 
 signedHeaders :: NormalisedHeaders -> SignedHeaders
 signedHeaders = Tag . BS8.intercalate ";" . map fst . untag
 
 normaliseHeaders :: [Header] -> NormalisedHeaders
 normaliseHeaders =
-  -- FIXME: convert this to an ordered map.
   Tag
     . map (first CI.foldedCase)
-    . List.nubBy ((==) `Function.on` fst)
-    . List.sortBy (compare `Function.on` fst)
-    . filter ((/= "authorization") . fst)
-    . filter ((/= "content-length") . fst)
+    . Map.toAscList
```

Comment:
The original and somewhat unfounded intuition was that when the list of headers is small, in the range of <10ish items, then you're paying more to construct/allocate the `Map`.  Due to the use of `nubBy` you're getting `O(n^2)` and possibly the moon, but it all depends on the size of the input as to how much of an effect this actually has in reality.

At any rate, the `Map` code is arguably nicer, so even in the absence of performance-related evidence I'd be fine with merging it.

Reply:
I can see there's some irony with what I've said above and also suggesting this, but I think the cleaner code is worth it.